src/test/java/com/snowflake/kafka/connector/internal/streaming/StreamingClientProviderTest.java

```diff
@@ -101,34 +106,66 @@ public void testFirstGetClient() {
     this.invalidClient = Mockito.mock(SnowflakeStreamingIngestClient.class);
     Mockito.when(this.invalidClient.isClosed()).thenReturn(true);
+
+    // inject new handler and cache
+    StreamingClientHandler injectedStreamingClientHandler =
+        Mockito.spy(StreamingClientHandler.class);
+    LoadingCache<Map<String, String>, SnowflakeStreamingIngestClient> injectedRegistrationClients =
+        Caffeine.newBuilder()
+            .maximumSize(Runtime.getRuntime().maxMemory())
+            .removalListener(
+                (Map<String, String> key,
+                    SnowflakeStreamingIngestClient client,
+                    RemovalCause removalCause) -> {
+                  injectedStreamingClientHandler.closeClient(client);
+                })
+            .build(injectedStreamingClientHandler::createClient);
+    injectedRegistrationClients.put(validClientConfig, validClient);
+    injectedRegistrationClients.put(invalidClientConfig, invalidClient);
+
     StreamingClientProvider injectedProvider =
-        getStreamingClientProviderForTests(this.invalidClient, this.streamingClientHandler);
+        getStreamingClientProviderForTests(
+            injectedStreamingClientHandler, injectedRegistrationClients);
 
-    // test: getting invalid client with valid config
+    // test: getting valid client
     this.validClient = injectedProvider.getClient(validClientConfig);
 
-    // verify: created valid client
+    // verify: valid client was got, but if optimization enabled we didnt need to create a new
+    // client
     assert StreamingClientHandler.isClientValid(this.validClient);
     assert this.validClient.getName().contains(validClientConfig.get(Utils.NAME));
     assert !this.validClient.getName().contains(invalidClientConfig.get(Utils.NAME));
-    Mockito.verify(this.streamingClientHandler, Mockito.times(1)).createClient(validClientConfig);
-
-    // verify: invalid client was closed, depending on optimization
-    Mockito.verify(this.invalidClient, Mockito.times(this.enableClientOptimization ? 1 : 0))
-        .isClosed();
+    Mockito.verify(
+            injectedStreamingClientHandler, Mockito.times(this.enableClientOptimization ? 0 : 1))
+        .createClient(validClientConfig);
+
+    // test: getting invalid client
+    this.invalidClient = injectedProvider.getClient(invalidClientConfig);
+
+    // verify: invalid client was refreshed / recreated
+    assert StreamingClientHandler.isClientValid(this.invalidClient);
+    assert !this.invalidClient.getName().contains(validClientConfig.get(Utils.NAME));
+    assert this.invalidClient.getName().contains(invalidClientConfig.get(Utils.NAME));
+    Mockito.verify(injectedStreamingClientHandler, Mockito.times(1))
+        .createClient(invalidClientConfig);
   }
 
   @Test
   public void testGetExistingClient() {
-    // test
+    // setup configs with different roles
+    this.clientConfig1.put(SF_ROLE, "testrole_kafka");
```

Comment:
lets not reveal our connector role, since this is UT we can use any random name right?

Reply:
no, this actually creates the client, we are only spy-ing on the streamingHandler so it creates the client. i'll see if i can use a different role

Reply:
Made the StreamingClientProviderTest file actually UT so we don't use the role anymore. Split off a separate StreamingClientProviderIT that will reveal the role though